In [1]:
%env HV_DOC_HTML=True

env: HV_DOC_HTML=True


In [2]:
pip install -q hvplot

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 155.4/155.4 kB 5.7 MB/s eta 0:00:00


In [3]:
pip install -q holoviews

In [4]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [5]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [6]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [7]:
# Plot your data to see what's in your DataFrame
hvplot.extension("bokeh")
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [8]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaledData = StandardScaler().fit_transform(df_market_data)

In [9]:
# Create a DataFrame with the scaled data
scaledMarketDataDF = pd.DataFrame(
    scaledData,
    columns=df_market_data.columns
)

# Copy the crypto names from the original data
scaledMarketDataDF['coin_id'] = df_market_data.index

# Set the coinid column as index
scaledMarketDataDF = scaledMarketDataDF.set_index("coin_id")

# Display sample data
scaledMarketDataDF.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [10]:
# Create a list with the number of k-values from 1 to 11
kValues = list(range(1,11))

In [11]:
# Create an empty list to store the inertia values
inertiaValues = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for k in kValues:
  kModel = KMeans(n_clusters=k, random_state=0, n_init='auto')
  kModel.fit(scaledMarketDataDF)
  inertiaValues.append(kModel.inertia_)

inertiaValues[:5]

[287.0,
 198.5718180979813,
 145.63574772192862,
 79.02243535120978,
 66.21778006000338]

In [12]:
# Create a dictionary with the data to plot the Elbow curve
elbowData = {
    'k':kValues,
    'inertia':inertiaValues
}

# Create a DataFrame with the data to plot the Elbow curve
elbowDF = pd.DataFrame(elbowData)

elbowDF.head()

k     inertia
0  1  287.000000
1  2  198.571818
2  3  145.635748
3  4   79.022435
4  5   66.217780

In [13]:
from re import X
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
hvplot.extension("bokeh")

elbowPlot = elbowDF.hvplot.line(
    x='k',
    y='inertia',
    title="Elbow Curve",
    xticks = kValues
)

elbowPlot

:Curve   [k]   (inertia)

#### Answer the following question:

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [14]:
# Initialize the K-Means model using the best value for k
kModel = KMeans(n_clusters=4, n_init='auto')

In [15]:
# Fit the K-Means model using the scaled data
kModel.fit(scaledMarketDataDF)

KMeans(n_clusters=4, n_init='auto')

In [16]:
# Predict the clusters to group the cryptocurrencies using the scaled data
cryptoPrediction = kModel.predict(scaledMarketDataDF)

# Print the resulting array of cluster values.
print(cryptoPrediction)

[3 3 0 0 3 3 3 3 3 0 0 0 0 3 0 3 0 0 3 0 0 3 0 0 0 0 0 0 3 0 0 0 2 3 0 0 1
 0 0 0 0]


In [17]:
# Create a copy of the DataFrame
scaledMarketDataPredictionsDF = scaledMarketDataDF.copy()

In [18]:
# Add a new column to the DataFrame with the predicted clusters
scaledMarketDataPredictionsDF['cryptoCluster'] = cryptoPrediction

# Display sample data
scaledMarketDataPredictionsDF.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  cryptoCluster  
coin_id                                                  
bitcoin                        -0.251637              3  
ethereum                       -0.199352              3  
tether                         -0.282061              0  
ripple                         -0.295546              0  
bitcoin-cash                   -0.270317              3

In [19]:
hvplot.extension("bokeh")
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
clusterPlot = scaledMarketDataPredictionsDF.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    title="Scatter Plot",
    by="cryptoCluster",
    hover_cols=["coin_id"],
    marker=["hex","square","cross","triangle"]
)

clusterPlot

:NdOverlay   [cryptoCluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [20]:
# Create a PCA model instance and set `n_components=3`.
pcaModel = PCA(n_components=3)

In [21]:
# Use the PCA model with `fit_transform` to reduce to
# three principal components.
cryptoPCAData = pcaModel.fit_transform(scaledMarketDataDF)

# View the first five rows of the DataFrame.
cryptoPCAData[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [22]:
pd.DataFrame(cryptoPCAData).head()

0         1         2
0 -0.600667  0.842760  0.461595
1 -0.458261  0.458466  0.952877
2 -0.433070 -0.168126 -0.641752
3 -0.471835 -0.222660 -0.479053
4 -1.157800  2.041209  1.859715

In [23]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pcaModel.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [25]:
sum(pcaModel.explained_variance_ratio_)

0.895031657030984

In [24]:
# Dump of all features that contributed to variance ratios of each PC
pd.DataFrame(
    pcaModel.components_,
    columns=scaledMarketDataDF.columns,
    index=['PC1','PC2','PC3']
)

price_change_percentage_24h  price_change_percentage_7d  \
PC1                    -0.416728                   -0.102432   
PC2                     0.358162                    0.227432   
PC3                    -0.218795                    0.787670   

     price_change_percentage_14d  price_change_percentage_30d  \
PC1                    -0.009162                     0.191523   
PC2                     0.540415                     0.562182   
PC3                     0.349534                    -0.181744   

     price_change_percentage_60d  price_change_percentage_200d  \
PC1                     0.320365                      0.594468   
PC2                     0.434040                      0.027735   
PC3                    -0.361377                      0.039733   

     price_change_percentage_1y  
PC1                    0.568379  
PC2                   -0.150789  
PC3                    0.210541

#### Answer the following question:

**Question:** What is the total explained variance of the three principal components?

**Answer:** f"The total explained variance: ~89.5%"

In [26]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data.
pcaDF = pd.DataFrame(
    cryptoPCAData,
    columns=['PC1','PC2','PC3']
)

# Copy the crypto names from the original data
pcaDF["coin_id"] = df_market_data.index

# Set the coinid column as index
pcaDF = pcaDF.set_index("coin_id")

# Display sample data
pcaDF.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [27]:
# Create a list with the number of k-values from 1 to 11
kValues = list(range(1,11))

In [28]:
# Create an empty list to store the inertia values
inertiaValues = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for k in kValues:
  kModel = KMeans(n_clusters=k, random_state=0, n_init="auto")
  kModel.fit(pcaDF)
  inertiaValues.append(kModel.inertia_)

inertiaValues[:5]

[256.8740855678924,
 168.81189630685216,
 115.9180766332006,
 49.66549665179734,
 41.35731380312854]

In [29]:
# Create a dictionary with the data to plot the Elbow curve
pcaElbow = {
    "k":kValues,
    "inertia":inertiaValues
}
# Create a DataFrame with the data to plot the Elbow curve
pcaElbowDF = pd.DataFrame(pcaElbow)
pcaElbowDF.head()

k     inertia
0  1  256.874086
1  2  168.811896
2  3  115.918077
3  4   49.665497
4  5   41.357314

In [30]:
hvplot.extension("bokeh")
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
pcaElbowPlot = pcaElbowDF.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve Using PCA Data",
    xticks=kValues
)

pcaElbowPlot

:Curve   [k]   (inertia)

#### Answer the following questions:

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, both were equal to '4'

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [31]:
# Initialize the K-Means model using the best value for k
pcaKModel = KMeans(n_clusters=4, n_init='auto')

In [32]:
# Fit the K-Means model using the PCA data
pcaKModel.fit(pcaDF)

KMeans(n_clusters=4, n_init='auto')

In [33]:
# Predict the clusters to group the cryptocurrencies using the PCA data
pcaPredictions = pcaKModel.predict(pcaDF)
# Print the resulting array of cluster values.
print(pcaPredictions)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 2 0 0 0 3
 0 0 0 0]


In [34]:
# Create a copy of the DataFrame with the PCA data
pcaPredictionsDF = pcaDF.copy()

# Add a new column to the DataFrame with the predicted clusters
pcaPredictionsDF['PCA Predicted Cluster'] = pcaPredictions

# Display sample data
pcaPredictionsDF.head()

PC1       PC2       PC3  PCA Predicted Cluster
coin_id                                                          
bitcoin      -0.600667  0.842760  0.461595                      0
ethereum     -0.458261  0.458466  0.952877                      0
tether       -0.433070 -0.168126 -0.641752                      0
ripple       -0.471835 -0.222660 -0.479053                      0
bitcoin-cash -1.157800  2.041209  1.859715                      0

In [35]:
hvplot.extension('bokeh')
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
clusterPCAPlot = pcaPredictionsDF.hvplot.scatter(
    x='PC1',
    y='PC2',
    by="PCA Predicted Cluster",
    marker=['hex','square','cross','triangle']
)

clusterPCAPlot

:NdOverlay   [PCA Predicted Cluster]
   :Scatter   [PC1]   (PC2)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [36]:
# Composite plot to contrast the Elbow curves
hvplot.extension('bokeh')
elbowPlot + pcaElbowPlot

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [37]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!
hvplot.extension('bokeh')
clusterPlot + clusterPCAPlot

:Layout
   .NdOverlay.I  :NdOverlay   [cryptoCluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [PCA Predicted Cluster]
      :Scatter   [PC1]   (PC2)

#### Answer the following question:

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Less features can be used to get a similar performance on the model since 4 clusters are clearly identified.